waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,248 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.oidc
+  (:require [clj-time.coerce :as tc]
+            [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [waiter.auth.authentication :as auth]
+            [waiter.auth.jwt :as jwt]
+            [waiter.cookie-support :as cookie-support]
+            [waiter.status-codes :refer :all]
+            [waiter.util.http-utils :as hu]
+            [waiter.util.ring-utils :as ru]
+            [waiter.util.utils :as utils])
+  (:import (java.net URI)))
+
+(def ^:const accept-redirect-header-name "accept-redirect")
+
+(def ^:const accept-redirect-auth-header-name "accept-redirect-auth")
+
+(def ^:const challenge-cookie-duration-secs 60)
+
+(def ^:const code-verifier-length 128)
+
+(def ^:const content-security-policy-value "default-src 'none'; frame-ancestors 'none'")
+
+(def ^:const oidc-challenge-cookie "x-waiter-oidc-challenge")
+
+(def ^:const oidc-callback-uri "/oidc/v1/callback")
+
+(defn create-code-verifier
+  "Creates a randomly generated string containing characters from A-Z of length code-verifier-length."
+  []
+  (apply str (take code-verifier-length (repeatedly #(char (+ (rand 26) 65))))))
```

Comment:
What's `rand`?  Is it like the C library's `rand(3)`?  My preference would be to get N random bytes using an API that clearly must be doing the right thing under the covers, then base64-encode that, which will get you ~64^N bits of entropy vs. ~52^N here where the N are not necessarily well-chosen.

Reply:
I changed to `rand-nth`, but it and `rand` use instances of `java.util.Random` which are not cryptographically secure.

I changed to a variant that uses `java.security.SecureRandom`which provides a cryptographically strong random number generator.